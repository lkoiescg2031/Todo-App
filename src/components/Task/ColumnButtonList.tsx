import React, { PureComponent } from "react";

import Button from "@material-ui/core/Button";

import { ColumnItem } from "../Column/Column";

interface ColumnButtonListProps {
	onColumnSelected: (selectedColumn: ColumnItem) => void;
	columnId: number;
	columns: ColumnItem[];
	classes?: {};
}

export default class ColumnButtonList extends PureComponent<ColumnButtonListProps> {
	constructor(props: ColumnButtonListProps) {
		super(props);

		this.state = {};
		this.handleOnClick = this.handleOnClick.bind(this);
	}

	handleOnClick(column: ColumnItem) {
		return () => {
			this.props.onColumnSelected(column);
		};
	}

	render() {
		const { columnId, columns, classes } = this.props;
		return columns
			.filter((column) => column.id !== columnId)
			.map((column, idx) => (
				<Button
					key={`columnMenu-${column.id}`}
					variant="contained"
					disableElevation
					classes={classes}
					onClick={this.handleOnClick(column)}
				>
					{`${column.name} 으로 이동`}
				</Button>
			));
	}
}
