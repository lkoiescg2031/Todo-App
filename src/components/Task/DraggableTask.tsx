import React from "react";
import { useDrag } from "react-dnd";

import Task, { TaskContainerProps, TaskItem } from "./Task";

export const TaskType = "task";

export interface DragItem {
	type: string;
	data: TaskItem;
}

const DraggableTask: React.FC<TaskContainerProps> = (props) => {
	const { data } = props;

	const [{ isDragging }, drag, dragPreviewRef] = useDrag(
		() => ({
			item: { type: TaskType, data },
			collect: (monitor) => ({
				isDragging: !!monitor.isDragging(),
			}),
		}),
		[data]
	);

	return (
		<div ref={drag} style={{ opacity: isDragging ? 0.5 : 1 }}>
			<Task dragPreviewRef={dragPreviewRef} {...props} />
		</div>
	);
};

export default DraggableTask;
