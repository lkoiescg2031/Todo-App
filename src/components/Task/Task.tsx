import React, { PureComponent } from "react";

import { DragElementWrapper, DragPreviewOptions } from "react-dnd";

import { withStyles } from "@material-ui/core/styles";

import Dialog from "@material-ui/core/Dialog";

import Card from "@material-ui/core/Card";
import CardActionArea from "@material-ui/core/CardActionArea";

import Typography from "@material-ui/core/Typography";
import TextField from "@material-ui/core/TextField";
import Button from "@material-ui/core/Button";

import ColumnButtonList from "../../container/ColumnButtonList";
import { ColumnItem } from "../Column/Column";

export interface TaskItem {
	id: number;
	name: string;
	columnId: number;
	boardId: number;
}

interface TaskProps {
	itemId: number;
	data: TaskItem;
}

export interface TaskContainerProps extends TaskProps {
	requestDeleteTask: (id: number, params?: {}, meta?: {}) => void;
	requestUpdateTask: (
		id: number,
		data: { name?: string; columnId?: number; boardId?: number },
		meta?: {}
	) => void;
}

interface DraggableTaskProps extends TaskContainerProps {
	dragPreviewRef: DragElementWrapper<DragPreviewOptions>;
}

interface TaskStylesProps extends DraggableTaskProps {
	classes: {
		root: string;
		title: string;
		dialogContainer: string;
		dialogPaper: string;
		dialogContent: string;
		dialogPaperScrollPaper: string;
		inputWrapper: string;
		textfieldRoot: string;
		inputRoot: string;
		saveButton: string;
		menuWrapper: string;
		deleteButton: string;
	};
}

interface TaskState {
	isOpen: boolean;
}

class Task extends PureComponent<TaskStylesProps, TaskState> {
	private cardRef: React.RefObject<HTMLDivElement>;
	private dialogInnerRef: React.RefObject<HTMLDivElement>;
	private textFieldRef: React.RefObject<HTMLDivElement>;
	private inputRef: React.RefObject<HTMLInputElement>;

	constructor(props: TaskStylesProps) {
		super(props);

		this.state = { isOpen: false };

		this.cardRef = React.createRef<HTMLDivElement>();
		this.dialogInnerRef = React.createRef<HTMLDivElement>();
		this.textFieldRef = React.createRef<HTMLDivElement>();
		this.inputRef = React.createRef<HTMLInputElement>();

		this.openDialog = this.openDialog.bind(this);
		this.closeDialog = this.closeDialog.bind(this);
		this.setupPosition = this.setupPosition.bind(this);
		this.handleUpdate = this.handleUpdate.bind(this);
		this.handleUpdateColumnId = this.handleUpdateColumnId.bind(this);
		this.handleDelete = this.handleDelete.bind(this);
		this.onRequestSuccess = this.onRequestSuccess.bind(this);
	}

	openDialog() {
		this.setState((prevState) => ({
			...prevState,
			isOpen: true,
		}));
	}

	closeDialog() {
		this.setState((prevState) => ({
			...prevState,
			isOpen: false,
		}));
	}

	setupPosition() {
		const cardEle = this.cardRef.current;
		const dialogInnerEle = this.dialogInnerRef.current;
		const textFieldEle = this.textFieldRef.current;
		const inputEle = this.inputRef.current;

		if (cardEle && dialogInnerEle && textFieldEle && inputEle) {
			const { top, left, width, height } = cardEle.getBoundingClientRect();
			dialogInnerEle.style.top = `${top}px`;
			dialogInnerEle.style.left = `${left}px`;
			textFieldEle.style.width = `${width}px`;
			inputEle.style.minHeight = `${Math.max(height, 80)}px`;
			inputEle.focus();
		}
	}

	handleUpdate() {
		const { itemId, data, requestUpdateTask } = this.props;
		const inputEle = this.inputRef.current;

		if (inputEle) {
			const name = inputEle.value;
			requestUpdateTask(itemId, { ...data, name });
			this.closeDialog();
		}
	}

	handleUpdateColumnId({ id: columnId }: ColumnItem) {
		const { itemId, data, requestUpdateTask } = this.props;
		requestUpdateTask(itemId, { ...data, columnId });
		this.closeDialog();
	}

	onRequestSuccess() {
		this.closeDialog();
	}

	handleDelete() {
		const { itemId, requestDeleteTask } = this.props;
		requestDeleteTask(itemId, {});
		this.closeDialog();
	}

	render() {
		const { data, dragPreviewRef, classes } = this.props;
		const { name: title, columnId } = data || {};
		const { isOpen } = this.state;

		return (
			<Card ref={this.cardRef} classes={{ root: classes.root }}>
				<CardActionArea onClick={this.openDialog}>
					<Typography
						ref={dragPreviewRef}
						classes={{ root: classes.title }}
						component="pre"
						variant="body1"
					>
						{title}
					</Typography>
				</CardActionArea>
				<Dialog
					classes={{
						container: classes.dialogContainer,
						paper: classes.dialogPaper,
						paperScrollPaper: classes.dialogPaperScrollPaper,
					}}
					open={isOpen}
					onClose={this.closeDialog}
					onEntering={this.setupPosition}
					innerRef={this.dialogInnerRef}
				>
					<div className={classes.dialogContent}>
						<div className={classes.inputWrapper}>
							<TextField
								ref={this.textFieldRef}
								color="primary"
								variant="outlined"
								size="small"
								classes={{ root: classes.inputRoot }}
								placeholder="Enter a title for this task..."
								defaultValue={title}
								multiline
								inputRef={this.inputRef}
								InputProps={{
									classes: {
										root: classes.textfieldRoot,
									},
								}}
							/>
							<Button
								variant="contained"
								color="primary"
								disableElevation
								classes={{ root: classes.saveButton }}
								onClick={this.handleUpdate}
							>
								Save
							</Button>
						</div>
						<div className={classes.menuWrapper}>
							<Button
								variant="contained"
								disableElevation
								classes={{ root: classes.deleteButton }}
								onClick={this.handleDelete}
							>
								Delete
							</Button>
							<ColumnButtonList
								classes={{ root: classes.deleteButton }}
								columnId={columnId}
								onColumnSelected={this.handleUpdateColumnId}
							/>
						</div>
					</div>
				</Dialog>
			</Card>
		);
	}
}

export default withStyles((theme) => ({
	root: {
		position: "relative",
	},
	title: {
		padding: theme.spacing(1),
	},
	dialogContainer: {
		position: "relative",
	},
	dialogPaper: {
		position: "absolute",
		top: 0,
		left: 0,
		backgroundColor: "#00000000",
		margin: 0,
		boxShadow: "none",
	},
	dialogContent: {
		display: "flex",
		flexDirection: "column-reverse",
		"& > :not(:first-child)": {
			marginBottom: theme.spacing(1),
			[theme.breakpoints.up("sm")]: {
				marginLeft: theme.spacing(1),
			},
		},
		[theme.breakpoints.up("sm")]: {
			flexDirection: "row",
		},
	},
	dialogPaperScrollPaper: {
		maxHeight: "none",
	},
	inputWrapper: {
		display: "flex",
		flexDirection: "column",
		"& > :not(:first-child)": {
			marginTop: theme.spacing(1),
		},
	},
	textfieldRoot: {
		backgroundColor: "#FFFFFF",
	},
	inputRoot: {
		width: "100%",
	},
	saveButton: {
		marginLeft: "auto",
		textTransform: "none",
	},
	menuWrapper: {
		display: "flex",
		flexDirection: "column-reverse",
		[theme.breakpoints.up("sm")]: {
			flexDirection: "column",
		},
		"& > :not(:first-child)": {
			marginBottom: theme.spacing(1),
			[theme.breakpoints.up("sm")]: {
				marginBottom: 0,
				marginTop: theme.spacing(1),
			},
		},
	},
	deleteButton: {
		width: "max-content",
	},
}))(Task);
