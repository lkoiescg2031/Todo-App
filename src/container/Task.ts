import { connect } from "react-redux";

import Task from "../components/Task/Task";

import {
	updateItem as requestUpdateCard,
	deleteItem as requestDeleteCard,
} from "../redux/actions/cardAction";

const mapStateToProps = (state: {}) => {
	return {};
};

const mapDispatchToProps = {
	requestUpdateCard,
	requestDeleteCard,
};

export default connect(mapStateToProps, mapDispatchToProps)(Task);
