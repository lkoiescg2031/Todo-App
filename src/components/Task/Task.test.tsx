import React from "react";
import { render } from "@testing-library/react";

import Task from "./Task";

describe("<Task />", () => {
	it("snapshot test", () => {
		const workName = "work1";
		const utils = render(<Task title={workName} />);

		expect(utils.container).toMatchSnapshot();
	});

	it("title Props Test", () => {
		const workName = "test work name";
		const utils = render(<Task title={workName} />);

		utils.getByText(workName);
	});
});
