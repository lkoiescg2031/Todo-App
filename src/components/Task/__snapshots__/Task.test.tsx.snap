// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<KanBanCard /> snapshot test 1`] = `
<div>
  <div
    class="MuiPaper-root MuiCard-root Card-root-1 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <button
      class="MuiButtonBase-root MuiCardActionArea-root Card-actionArea-2"
      tabindex="0"
      type="button"
    >
      <p
        class="MuiTypography-root MuiTypography-body1"
      />
      <span
        class="MuiCardActionArea-focusHighlight"
      />
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
